src/Uno.UI.Runtime.Skia.Gtk/Extensions/GtkNativeElementHostingExtension.cs

```diff
@@ -1,100 +1,200 @@
+internal class GtkNativeElementHostingExtension : ContentPresenter.INativeElementHostingExtension
 {
-	public GtkNativeElementHostingExtension()
+	private readonly ContentPresenter _presenter;
+
+	public GtkNativeElementHostingExtension(ContentPresenter contentPresenter)
 	{
+		_presenter = contentPresenter;
 	}
 
-	internal static Fixed? GetOverlayLayer(XamlRoot xamlRoot) =>
-		GtkManager.XamlRootMap.GetHostForRoot(xamlRoot)?.NativeOverlayLayer;
+	private static Dictionary<object, IDisposable> NativeRenderDisposables { get; } = new();
+
+	private XamlRoot? XamlRoot => _presenter.XamlRoot;
+	private Fixed? OverlayLayer => _presenter.XamlRoot is { } xamlRoot ? GtkManager.XamlRootMap.GetHostForRoot(xamlRoot)?.NativeOverlayLayer : null;
 
-	public bool IsNativeElement(object content)
-		=> content is Widget;
+	public bool IsNativeElement(object content) => content is Widget;
 
-	public void AttachNativeElement(object owner, object content)
+	public void AttachNativeElement(object content)
 	{
-		if (content is Widget widget
-			&& owner is XamlRoot xamlRoot
-			&& GetOverlayLayer(xamlRoot) is { } overlay)
+		if (content is Widget widget)
 		{
 			widget.ShowAll();
-			overlay.Put(widget, 0, 0);
+			// We don't attach the widget to the overlay here, but on the first arrange, to prevent the overlay from being
+			// misplaced for a split-second before the arrange occurs
 		}
 		else
 		{
 			if (this.Log().IsEnabled(LogLevel.Debug))
 			{
-				this.Log().Debug($"Unable to attach native element {content} to {owner}.");
+				this.Log().Debug($"Unable to attach native element {content} to {XamlRoot}.");
 			}
 		}
 	}
 
-	public void DetachNativeElement(object owner, object content)
+	public void DetachNativeElement(object content)
 	{
-		if (content is Widget widget
-			&& owner is XamlRoot xamlRoot
-			&& GetOverlayLayer(xamlRoot) is { } overlay)
+		if (content is Widget widget && OverlayLayer is { } overlay)
 		{
+			if (NativeRenderDisposables.Remove(widget, out var disposable))
+			{
+				disposable.Dispose();
+			}
 			overlay.Remove(widget);
 		}
 		else
 		{
 			if (this.Log().IsEnabled(LogLevel.Debug))
 			{
-				this.Log().LogDebug($"Unable to detach native element {content} from {owner}.");
+				this.Log().LogDebug($"Unable to detach native element {content} from {XamlRoot}.");
 			}
 		}
 	}
 
-	public bool IsNativeElementAttached(object owner, object nativeElement) =>
-		nativeElement is Widget widget
-			&& owner is XamlRoot xamlRoot
-			&& GetOverlayLayer(xamlRoot) is { } overlay
-			&& widget.Parent == overlay;
+	public object CreateSampleComponent(string text)
+	{
+		var vbox = new VBox(false, 5);
+
+		var label = new Label(text);
+		vbox.PackStart(label, false, false, 0);
+
+		var hbox = new HBox(true, 3);
+
+		var button1 = new Button("Button 1");
+		var button2 = new Button("Button 2");
+		hbox.Add(button1);
+		hbox.Add(button2);
+
+		vbox.PackStart(hbox, false, false, 0);
+
+		return vbox;
+	}
```

Comment:
Test code that should be removed? If not, let's move to a separate class.

Reply:
It's "test code" but we need to keep it for debugging. I would much rather keep everything tight in one class instead of creating a new interface and registering it with `ApiExtensibility` (especially since for X11, this method needs the added context of the specific ContentPresenter). I moved these methods to a separate partial for now.

Reply:
Separate partial is fine. Thanks.